chainforge/react-server/src/backend/query.ts

```diff
@@ -129,12 +130,13 @@ export class PromptPipeline {
    *                                 and 3 different prior chat histories, it will send off 9 queries. 
    * @yields Yields `LLMResponseObject` if API call succeeds, or `LLMResponseError` if API call fails, for all requests. 
    */
-  async *gen_responses(vars: Dict, 
+  async *gen_responses( id: string,
```

Comment:
So I'm not entirely convinced that the stop checks need to occur at this level of the code, and that is because we still always need to wait for `_prompt_llm` to return before we cancel. To keep things clearer, I would move all `id` logic out of `gen_responses` (a function should remain separate from the front-end as much as possible) and then break out of the for loop in line 674 `for await (const response of prompter.gen_responses(...`. So you'd check for the stop `id` within that body, in `queryllm`, but not pass the `id` further down into gen_responses. Does this make sense?

Reply:
I’ve changed my mind —keep the spirit of what you have (raising errors in _prompt_llm etc) but instead of “id” do this:

Add an optional input “checkStopCondition” to the end of gen_responses, instead of id. The checkStopCondition would be a user-passed boolean function. Just call that function to check if its true, if it is, raise the UserForcedPrematureExit. This way we still separate concerns between front and backend.